File: packages/core-rs/src/assert.rs
//! Assertion expression language per ARKY-KERNEL-v1 §4 / ARKY-ASSERTIONS-v1.
//!
//! Tri-valued (Kleene) result: Pass / Fail / Indeterminate. Grammar (§4.1):
//!   Expr        := Comparison | LogicalExpr | "(" Expr ")"
//!   Comparison  := Symbol Op Value | Symbol "in" "[" ValueList "]"
//!   LogicalExpr := Expr ("&&" | "||") Expr | "!" Expr
//!   Op          := < <= > >= == !=
//!
//! Mirrors @arky/core (TS) assert.ts; the two must agree on every K1 vector.

use std::cmp::Ordering;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TriState {
    Pass,
    Fail,
    Indeterminate,
}

impl TriState {
    pub fn as_str(self) -> &'static str {
        match self {
            TriState::Pass => "PASS",
            TriState::Fail => "FAIL",
            TriState::Indeterminate => "INDETERMINATE",
        }
    }
}

/// A bound symbol value (from TIM measurement.value).
#[derive(Debug, Clone, PartialEq)]
pub enum SymVal {
    Num(f64),
    Str(String),
    Bool(bool),
}

/// symbol name -> value (absent => Indeterminate).
pub type Symbols = std::collections::BTreeMap<String, SymVal>;

#[derive(Debug, Clone)]
pub struct EvalResult {
    pub result: TriState,
    pub error: Option<String>,
}

// --- tokens ---

#[derive(Debug, Clone, PartialEq)]
enum Tok {
    Num(f64),
    Str(String),
    Bool(bool),
    Sym(String),
    Op(String),
    LParen,
    RParen,
    LBrack,
    RBrack,
    Comma,
    And,
    Or,
    Not,
    In,
}

fn tokenize(src: &str) -> Result<Vec<Tok>, String> {
    let chars: Vec<char> = src.chars().collect();
    let mut toks = Vec::new();
    let mut i = 0;
    while i < chars.len() {
        let c = chars[i];
        match c {
            ' ' | '\t' => i += 1,
            '(' => {
                toks.push(Tok::LParen);
                i += 1;
            }
            ')' => {
                toks.push(Tok::RParen);
                i += 1;
            }
            '[' => {
                toks.push(Tok::LBrack);
                i += 1;
            }
            ']' => {
                toks.push(Tok::RBrack);
                i += 1;
            }
            ',' => {
                toks.push(Tok::Comma);
                i += 1;
            }
            '&' if i + 1 < chars.len() && chars[i + 1] == '&' => {
                toks.push(Tok::And);
                i += 2;
            }
            '|' if i + 1 < chars.len() && chars[i + 1] == '|' => {
                toks.push(Tok::Or);
                i += 2;
            }
            '!' if !(i + 1 < chars.len() && chars[i + 1] == '=') => {
                toks.push(Tok::Not);
                i += 1;
            }
            '"' => {
                let mut j = i + 1;
                let mut s = String::new();
                while j < chars.len() && chars[j] != '"' {
                    s.push(chars[j]);
                    j += 1;
                }
                if j >= chars.len() {
                    return Err("unterminated string".into());
                }
                toks.push(Tok::Str(s));
                i = j + 1;
            }
            '<' | '>' | '=' | '!' => {
                if i + 1 < chars.len() && chars[i + 1] == '=' {
                    toks.push(Tok::Op(format!("{}=", c)));
                    i += 2;
                } else if c == '<' || c == '>' {
                    toks.push(Tok::Op(c.to_string()));
                    i += 1;
                } else {
                    return Err(format!("bad operator at '{}'", c));
                }
            }
            '0'..='9' => {
                let mut j = i;
                let mut n = String::new();
                while j < chars.len() && (chars[j].is_ascii_digit() || chars[j] == '.') {
                    n.push(chars[j]);
                    j += 1;
                }
                toks.push(Tok::Num(n.parse().map_err(|_| "bad number")?));
                i = j;
            }
            c if c == '_' || c.is_ascii_lowercase() => {
                let mut j = i;
                let mut s = String::new();
                while j < chars.len() && (chars[j] == '_' || chars[j].is_ascii_lowercase() || chars[j].is_ascii_digit()) {
                    s.push(chars[j]);
                    j += 1;
                }
                match s.as_str() {
                    "true" => toks.push(Tok::Bool(true)),
                    "false" => toks.push(Tok::Bool(false)),
                    "in" => toks.push(Tok::In),
                    _ => toks.push(Tok::Sym(s)),
                }
                i = j;
            }
            other => return Err(format!("unexpected character '{}'", other)),
        }
    }
    Ok(toks)
}

// --- AST ---

#[derive(Debug, Clone)]
enum Lit {
    Num(f64),
    Str(String),
    Bool(bool),
}

#[derive(Debug, Clone)]
enum Ast {
    Cmp { sym: String, op: String, val: Lit },
    In { sym: String, vals: Vec<Lit> },
    And(Box<Ast>, Box<Ast>),
    Or(Box<Ast>, Box<Ast>),
    Not(Box<Ast>),
    SymRef(String),
}

struct Parser {
    toks: Vec<Tok>,
    pos: usize,
}

impl Parser {
    fn new(toks: Vec<Tok>) -> Self {
        Parser { toks, pos: 0 }
    }
    fn peek(&self) -> Option<&Tok> {
        self.toks.get(self.pos)
    }
    fn next(&mut self) -> Option<Tok> {
        let t = self.toks.get(self.pos).cloned();
        self.pos += 1;
        t
    }
    fn parse(&mut self) -> Result<Ast, String> {
        let e = self.parse_or()?;
        if self.pos != self.toks.len() {
            return Err("trailing tokens".into());
        }
        Ok(e)
    }
    fn parse_or(&mut self) -> Result<Ast, String> {
        let mut l = self.parse_and()?;
        while self.peek() == Some(&Tok::Or) {
            self.next();
            l = Ast::Or(Box::new(l), Box::new(self.parse_and()?));
        }
        Ok(l)
    }
    fn parse_and(&mut self) -> Result<Ast, String> {
        let mut l = self.parse_unary()?;
        while self.peek() == Some(&Tok::And) {
            self.next();
            l = Ast::And(Box::new(l), Box::new(self.parse_unary()?));
        }
        Ok(l)
    }
    fn parse_unary(&mut self) -> Result<Ast, String> {
        if self.peek() == Some(&Tok::Not) {
            self.next();
            return Ok(Ast::Not(Box::new(self.parse_unary()?)));
        }
        self.parse_primary()
    }
    fn parse_primary(&mut self) -> Result<Ast, String> {
        match self.peek().cloned() {
            Some(Tok::LParen) => {
                self.next();
                let e = self.parse_or()?;
                if self.next() != Some(Tok::RParen) {
                    return Err("expected )".into());
                }
                Ok(e)
            }
            Some(Tok::Sym(name)) => {
                self.next();
                match self.peek().cloned() {
                    Some(Tok::Op(op)) => {
                        self.next();
                        Ok(Ast::Cmp { sym: name, op, val: self.parse_lit()? })
                    }
                    Some(Tok::In) => {
                        self.next();
                        if self.next() != Some(Tok::LBrack) {
                            return Err("expected [".into());
                        }
                        let mut vals = vec![self.parse_lit()?];
                        while self.peek() == Some(&Tok::Comma) {
                            self.next();
                            vals.push(self.parse_lit()?);
                        }
                        if self.next() != Some(Tok::RBrack) {
                            return Err("expected ]".into());
                        }
                        Ok(Ast::In { sym: name, vals })
                    }
                    _ => Ok(Ast::SymRef(name)),
                }
            }
            _ => Err("expected symbol or (".into()),
        }
    }
    fn parse_lit(&mut self) -> Result<Lit, String> {
        match self.next() {
            Some(Tok::Num(n)) => Ok(Lit::Num(n)),
            Some(Tok::Str(s)) => Ok(Lit::Str(s)),
            Some(Tok::Bool(b)) => Ok(Lit::Bool(b)),
            _ => Err("expected literal".into()),
        }
    }
}

// --- Kleene logic ---

fn and3(a: TriState, b: TriState) -> TriState {
    use TriState::*;
    if a == Fail || b == Fail {
        Fail
    } else if a == Pass && b == Pass {
        Pass
    } else {
        Indeterminate
    }
}
fn or3(a: TriState, b: TriState) -> TriState {
    use TriState::*;
    if a == Pass || b == Pass {
        Pass
    } else if a == Fail && b == Fail {
        Fail
    } else {
        Indeterminate
    }
}
fn not3(a: TriState) -> TriState {
    match a {
        TriState::Pass => TriState::Fail,
        TriState::Fail => TriState::Pass,
        TriState::Indeterminate => TriState::Indeterminate,
    }
}
fn b(v: bool) -> TriState {
    if v {
        TriState::Pass
    } else {
        TriState::Fail
    }
}

fn eval_ast(ast: &Ast, symbols: &Symbols, errs: &mut Vec<String>) -> TriState {
    match ast {
        Ast::And(l, r) => and3(eval_ast(l, symbols, errs), eval_ast(r, symbols, errs)),
        Ast::Or(l, r) => or3(eval_ast(l, symbols, errs), eval_ast(r, symbols, errs)),
        Ast::Not(e) => not3(eval_ast(e, symbols, errs)),
        Ast::SymRef(sym) => match symbols.get(sym) {
            None => {
                errs.push(format!("no matching receipts for symbol '{}'", sym));
                TriState::Indeterminate
            }
            Some(SymVal::Bool(v)) => b(*v),
            Some(_) => {
                errs.push(format!("symbol '{}' is not boolean", sym));
                TriState::Indeterminate
            }
        },
        Ast::In { sym, vals } => match symbols.get(sym) {
            None => {
                errs.push(format!("no matching receipts for symbol '{}'", sym));
                TriState::Indeterminate
            }
            Some(v) => b(vals.iter().any(|l| lit_eq(l, v))),
        },
        Ast::Cmp { sym, op, val } => match symbols.get(sym) {
            None => {
                errs.push(format!("no matching receipts for symbol '{}'", sym));
                TriState::Indeterminate
            }
            Some(v) => eval_cmp(v, op, val, errs),
        },
    }
}

fn lit_eq(l: &Lit, v: &SymVal) -> bool {
    match (l, v) {
        (Lit::Num(a), SymVal::Num(b)) => a == b,
        (Lit::Str(a), SymVal::Str(b)) => a == b,
        (Lit::Bool(a), SymVal::Bool(b)) => a == b,
        _ => false,
    }
}

fn eval_cmp(v: &SymVal, op: &str, lit: &Lit, errs: &mut Vec<String>) -> TriState {
    // Type compatibility (§4.1).
    match (v, lit) {
        (SymVal::Num(_), Lit::Str(_)) => {
            errs.push("type mismatch: numeric symbol compared to string literal".into());
            return TriState::Indeterminate;
        }
        (SymVal::Num(_), Lit::Bool(_)) => {
            errs.push("type mismatch: numeric symbol compared to boolean literal".into());
            return TriState::Indeterminate;
        }
        (SymVal::Str(_), l) if !matches!(l, Lit::Str(_)) && op != "==" && op != "!=" => {
            errs.push("type mismatch: string symbol compared to non-string literal".into());
            return TriState::Indeterminate;
        }
        _ => {}
    }
    b(compare(v, op, lit))
}

fn compare(v: &SymVal, op: &str, lit: &Lit) -> bool {
    let ord = match (v, lit) {
        (SymVal::Num(a), Lit::Num(b)) => a.partial_cmp(b),
        (SymVal::Str(a), Lit::Str(b)) => Some(a.cmp(b)),
        (SymVal::Bool(a), Lit::Bool(b)) => Some(a.cmp(b)),
        _ => None,
    };
    match op {
        "==" => matches!(ord, Some(Ordering::Equal)),
        "!=" => !matches!(ord, Some(Ordering::Equal)),
        "<" => matches!(ord, Some(Ordering::Less)),
        "<=" => matches!(ord, Some(Ordering::Less) | Some(Ordering::Equal)),
        ">" => matches!(ord, Some(Ordering::Greater)),
        ">=" => matches!(ord, Some(Ordering::Greater) | Some(Ordering::Equal)),
        _ => false,
    }
}

/// Parse + evaluate an assertion string to a tri-state result.
pub fn evaluate_assertion(expr: &str, symbols: &Symbols) -> EvalResult {
    let ast = match tokenize(expr).and_then(|t| Parser::new(t).parse()) {
        Ok(a) => a,
        Err(e) => {
            return EvalResult {
                result: TriState::Indeterminate,
                error: Some(format!("parse error: {}", e)),
            }
        }
    };
    let mut errs = Vec::new();
    let result = eval_ast(&ast, symbols, &mut errs);
    EvalResult {
        result,
        error: if result == TriState::Indeterminate && !errs.is_empty() {
            Some(errs.remove(0))
        } else {
            None
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn syms() -> Symbols {
        let mut m = Symbols::new();
        m.insert("temp".into(), SymVal::Num(22.5));
        m.insert("flag".into(), SymVal::Bool(true));
        m
    }

    #[test]
    fn comparisons_and_logic() {
        let s = syms();
        assert_eq!(evaluate_assertion("temp > 20", &s).result, TriState::Pass);
        assert_eq!(evaluate_assertion("temp > 30", &s).result, TriState::Fail);
        assert_eq!(evaluate_assertion("temp >= 20 && temp <= 25", &s).result, TriState::Pass);
        assert_eq!(evaluate_assertion("temp > 30 || temp < 25", &s).result, TriState::Pass);
        assert_eq!(evaluate_assertion("!(temp > 30)", &s).result, TriState::Pass);
        assert_eq!(evaluate_assertion("temp in [22.5, 30]", &s).result, TriState::Pass);
        assert_eq!(evaluate_assertion("flag", &s).result, TriState::Pass);
    }

    #[test]
    fn missing_and_type_mismatch() {
        let s = syms();
        assert_eq!(evaluate_assertion("humidity > 50", &s).result, TriState::Indeterminate);
        let r = evaluate_assertion("temp > \"high\"", &s);
        assert_eq!(r.result, TriState::Indeterminate);
        assert!(r.error.unwrap().contains("type mismatch"));
    }
}

File: packages/core/src/assert.ts
/**
 * Assertion expression language per ARKY-KERNEL-v1 §4 / ARKY-ASSERTIONS-v1.
 *
 * Tri-valued (Kleene) result: PASS (true), FAIL (false), INDETERMINATE (cannot
 * evaluate — missing symbol, type/unit mismatch). Grammar (§4.1):
 *   Expr       := Comparison | LogicalExpr | "(" Expr ")"
 *   Comparison := Symbol Op Value | Symbol "in" "[" ValueList "]"
 *   LogicalExpr:= Expr ("&&" | "||") Expr | "!" Expr
 *   Op         := < <= > >= == !=
 */

export type TriState = 'PASS' | 'FAIL' | 'INDETERMINATE';

export interface Symbols {
  /** symbol name -> { value, unit? }; absent symbol => INDETERMINATE */
  [name: string]: { value: number | string | boolean; unit?: string } | undefined;
}

export interface EvalResult {
  result: TriState;
  error?: string;
}

// --- tokenizer ---

type Tok =
  | { t: 'num'; v: number }
  | { t: 'str'; v: string }
  | { t: 'bool'; v: boolean }
  | { t: 'sym'; v: string }
  | { t: 'op'; v: string }
  | { t: 'lparen' }
  | { t: 'rparen' }
  | { t: 'lbrack' }
  | { t: 'rbrack' }
  | { t: 'comma' }
  | { t: 'and' }
  | { t: 'or' }
  | { t: 'not' }
  | { t: 'in' };

function tokenize(src: string): Tok[] {
  const toks: Tok[] = [];
  let i = 0;
  const isSymStart = (c: string) => /[a-z_]/.test(c);
  const isSym = (c: string) => /[a-z0-9_]/.test(c);
  while (i < src.length) {
    const c = src[i];
    if (c === ' ' || c === '\t') {
      i++;
      continue;
    }
    if (c === '(') {
      toks.push({ t: 'lparen' });
      i++;
      continue;
    }
    if (c === ')') {
      toks.push({ t: 'rparen' });
      i++;
      continue;
    }
    if (c === '[') {
      toks.push({ t: 'lbrack' });
      i++;
      continue;
    }
    if (c === ']') {
      toks.push({ t: 'rbrack' });
      i++;
      continue;
    }
    if (c === ',') {
      toks.push({ t: 'comma' });
      i++;
      continue;
    }
    if (c === '&' && src[i + 1] === '&') {
      toks.push({ t: 'and' });
      i += 2;
      continue;
    }
    if (c === '|' && src[i + 1] === '|') {
      toks.push({ t: 'or' });
      i += 2;
      continue;
    }
    if (c === '!' && src[i + 1] !== '=') {
      toks.push({ t: 'not' });
      i++;
      continue;
    }
    if (c === '"') {
      let j = i + 1;
      let s = '';
      while (j < src.length && src[j] !== '"') s += src[j++];
      if (src[j] !== '"') throw new SyntaxError('unterminated string');
      toks.push({ t: 'str', v: s });
      i = j + 1;
      continue;
    }
    if (/[<>=!]/.test(c)) {
      const two = src.slice(i, i + 2);
      if (['<=', '>=', '==', '!='].includes(two)) {
        toks.push({ t: 'op', v: two });
        i += 2;
        continue;
      }
      if (c === '<' || c === '>') {
        toks.push({ t: 'op', v: c });
        i++;
        continue;
      }
      throw new SyntaxError(`bad operator near '${src.slice(i)}'`);
    }
    if (/[0-9]/.test(c)) {
      let j = i;
      let n = '';
      while (j < src.length && /[0-9.]/.test(src[j])) n += src[j++];
      toks.push({ t: 'num', v: Number(n) });
      i = j;
      continue;
    }
    if (isSymStart(c)) {
      let j = i;
      let s = '';
      while (j < src.length && isSym(src[j])) s += src[j++];
      if (s === 'true') toks.push({ t: 'bool', v: true });
      else if (s === 'false') toks.push({ t: 'bool', v: false });
      else if (s === 'in') toks.push({ t: 'in' });
      else toks.push({ t: 'sym', v: s });
      i = j;
      continue;
    }
    throw new SyntaxError(`unexpected character '${c}'`);
  }
  return toks;
}

// --- parser -> AST ---

type Lit = { k: 'num'; v: number } | { k: 'str'; v: string } | { k: 'bool'; v: boolean };
type Ast =
  | { k: 'cmp'; sym: string; op: string; val: Lit }
  | { k: 'in'; sym: string; vals: Lit[] }
  | { k: 'and'; l: Ast; r: Ast }
  | { k: 'or'; l: Ast; r: Ast }
  | { k: 'not'; e: Ast }
  | { k: 'symref'; sym: string }; // bare boolean symbol (e.g. manual_override)

class Parser {
  constructor(
    private toks: Tok[],
    private pos = 0,
  ) {}
  peek(): Tok | undefined {
    return this.toks[this.pos];
  }
  next(): Tok {
    return this.toks[this.pos++];
  }

  parse(): Ast {
    const e = this.parseOr();
    if (this.pos !== this.toks.length) throw new SyntaxError('trailing tokens');
    return e;
  }
  parseOr(): Ast {
    let l = this.parseAnd();
    while (this.peek()?.t === 'or') {
      this.next();
      l = { k: 'or', l, r: this.parseAnd() };
    }
    return l;
  }
  parseAnd(): Ast {
    let l = this.parseUnary();
    while (this.peek()?.t === 'and') {
      this.next();
      l = { k: 'and', l, r: this.parseUnary() };
    }
    return l;
  }
  parseUnary(): Ast {
    if (this.peek()?.t === 'not') {
      this.next();
      return { k: 'not', e: this.parseUnary() };
    }
    return this.parsePrimary();
  }
  parsePrimary(): Ast {
    const tk = this.peek();
    if (tk?.t === 'lparen') {
      this.next();
      const e = this.parseOr();
      if (this.next().t !== 'rparen') throw new SyntaxError('expected )');
      return e;
    }
    if (tk?.t === 'sym') {
      this.next();
      const nx = this.peek();
      if (nx?.t === 'op') {
        const op = (this.next() as any).v;
        return { k: 'cmp', sym: tk.v, op, val: this.parseLit() };
      }
      if (nx?.t === 'in') {
        this.next();
        if (this.next().t !== 'lbrack') throw new SyntaxError('expected [');
        const vals: Lit[] = [this.parseLit()];
        while (this.peek()?.t === 'comma') {
          this.next();
          vals.push(this.parseLit());
        }
        if (this.next().t !== 'rbrack') throw new SyntaxError('expected ]');
        return { k: 'in', sym: tk.v, vals };
      }
      return { k: 'symref', sym: tk.v }; // bare boolean symbol
    }
    throw new SyntaxError('expected symbol or (');
  }
  parseLit(): Lit {
    const tk = this.next();
    if (tk.t === 'num') return { k: 'num', v: tk.v };
    if (tk.t === 'str') return { k: 'str', v: tk.v };
    if (tk.t === 'bool') return { k: 'bool', v: tk.v };
    throw new SyntaxError('expected literal');
  }
}

// --- Kleene logic over tri-state ---

function and3(a: TriState, b: TriState): TriState {
  if (a === 'FAIL' || b === 'FAIL') return 'FAIL';
  if (a === 'PASS' && b === 'PASS') return 'PASS';
  return 'INDETERMINATE';
}
function or3(a: TriState, b: TriState): TriState {
  if (a === 'PASS' || b === 'PASS') return 'PASS';
  if (a === 'FAIL' && b === 'FAIL') return 'FAIL';
  return 'INDETERMINATE';
}
function not3(a: TriState): TriState {
  if (a === 'PASS') return 'FAIL';
  if (a === 'FAIL') return 'PASS';
  return 'INDETERMINATE';
}
const bool = (b: boolean): TriState => (b ? 'PASS' : 'FAIL');

/** Evaluate a parsed assertion against bound symbols. Captures the first error. */
function evalAst(ast: Ast, symbols: Symbols, errs: string[]): TriState {
  switch (ast.k) {
    case 'and':
      return and3(evalAst(ast.l, symbols, errs), evalAst(ast.r, symbols, errs));
    case 'or':
      return or3(evalAst(ast.l, symbols, errs), evalAst(ast.r, symbols, errs));
    case 'not':
      return not3(evalAst(ast.e, symbols, errs));
    case 'symref': {
      const s = symbols[ast.sym];
      if (s === undefined) {
        errs.push(`no matching receipts for symbol '${ast.sym}'`);
        return 'INDETERMINATE';
      }
      if (typeof s.value === 'boolean') return bool(s.value);
      errs.push(`symbol '${ast.sym}' is not boolean`);
      return 'INDETERMINATE';
    }
    case 'in': {
      const s = symbols[ast.sym];
      if (s === undefined) {
        errs.push(`no matching receipts for symbol '${ast.sym}'`);
        return 'INDETERMINATE';
      }
      const match = ast.vals.some((l) => l.v === s.value);
      return bool(match);
    }
    case 'cmp': {
      const s = symbols[ast.sym];
      if (s === undefined) {
        errs.push(`no matching receipts for symbol '${ast.sym}'`);
        return 'INDETERMINATE';
      }
      const lit = ast.val;
      // Type compatibility (§4.1 coercion rules).
      if (typeof s.value === 'number' && lit.k !== 'num') {
        errs.push(
          'type mismatch: numeric symbol compared to ' +
            (lit.k === 'str' ? 'string' : 'boolean') +
            ' literal',
        );
        return 'INDETERMINATE';
      }
      if (typeof s.value === 'string' && lit.k !== 'str') {
        if (ast.op === '==' || ast.op === '!=') {
          /* allow */
        } else {
          errs.push('type mismatch: string symbol compared to non-string literal');
          return 'INDETERMINATE';
        }
      }
      return bool(compare(s.value, ast.op, lit.v));
    }
  }
}

function compare(a: unknown, op: string, b: unknown): boolean {
  switch (op) {
    case '==':
      return a === b;
    case '!=':
      return a !== b;
    case '<':
      return (a as number) < (b as number);
    case '<=':
      return (a as number) <= (b as number);
    case '>':
      return (a as number) > (b as number);
    case '>=':
      return (a as number) >= (b as number);
    default:
      throw new Error(`unknown op ${op}`);
  }
}

/** Parse + evaluate an assertion string to a tri-state result. */
export function evaluateAssertion(expr: string, symbols: Symbols): EvalResult {
  let ast: Ast;
  try {
    ast = new Parser(tokenize(expr)).parse();
  } catch (e) {
    return { result: 'INDETERMINATE', error: `parse error: ${(e as Error).message}` };
  }
  const errs: string[] = [];
  const result = evalAst(ast, symbols, errs);
  return { result, ...(result === 'INDETERMINATE' && errs.length ? { error: errs[0] } : {}) };
}
